### ISSUE TITLE
Consider implementing JoinHandle::try_join

### ISSUE THREAD

**Issue by andreastedile** [tokio-rs/tokio#6488]
Consider the following code using a `JoinHandle`:

```rust
if handle.is_finished() {
    let result = handle.await;
    match result {
        Ok(result) => {}
        Err(error) => {}
    }
}
```

I think that it would a bit more ergonomic to be able to just do:

```rust
match handle.try_join() {
    Some(Ok(result)) => {} // Task has completed successfully
    Some(Err(error)) => {} // Task has completed with error
    None => {} // Task has not completed yet
}
```

As pointed out on Discord, there is the issue that that `handle.await` will panic if called after `handle.try_join` returns not `None`.

**Comment by dd-dreams** [comment]
> there is the issue that that handle.await will panic if called after handle.try_join returns not None
Though you should call `handle.try_join().await`. Why would you call `handle.await` after `try_join`?

**Comment by Imberflur** [comment]
This is presented here in an async context. I found this while looking for the ideal way to poll a `JoinHandle` from sync code. IMO this would be convenient for that case. Similar to some of the use cases for https://docs.rs/tokio/1.39.2/tokio/sync/oneshot/struct.Receiver.html#method.try_recv.

**Comment by Darksonn** [comment]
I suppose this could make sense to have, though the footgun about `.await` is unfortunate.

**Comment by Imberflur** [comment]
> the footgun about `.await` is unfortunate

Yeah... FWIW I don't have a case where this is essential, it would just be more convenient. So I could see it not being worth the additional footgun. At least for my particular case.

I.e. I could either pass results through a oneshot channel, check `is_finished` and hope that ensures [`FutureExt::now_or_never`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never) succeeds, or just manually poll the future with a dummy waker.
